Infer FnSig via Fn traits

adamrk opened the pull request:
Addresses https://github.com/rust-analyzer/rust-analyzer/issues/4481.
When inferring types check if the callee implements one of the builtin `Fn` traits. Also autoderef the callee before trying to figure out it's `FnSig`.

flodiebold commented on crates/ra_hir_ty/src/infer/expr.rs:
```diff
@@ -63,6 +63,75 @@ impl<'a> InferenceContext<'a> {
+                            };
+                            let return_ty = self.normalize_projection_ty(output_proj_ty);
+                            return Some(FnSig::from_params_and_return(
+                                parameters.into_iter().map(|ty| ty.clone()).collect(),
```
these parameters are still canonicalized. (I.e. there might be parameters that can be anything, which are represented by bound variables in the canonicalized solution and would need to be turned into inference variables.) Fortunately, that will resolve itself with the refactorings I mention above, I think.

adamrk replied:
I'm not following how this is resolved by the other refactorings. Should we be passing in the argument expressions for the call and somehow registering the new obligations without ever unpacking the types from the solution?

flodiebold replied:
What I mean is basically:
 - callable_sig gets passed the number of arguments (let's say 3)
 - callable_sig creates that many type variables (let's say `?1`, `?2`, `?3`)
 - callable_sig checks that the callee may implement `Fn<(?1, ?2, ?3)>`, and if so, registers that as an obligation and returns these variables `?1`, `?2`, `?3` as argument types.

Then what happens automatically is that that obligation gets checked later, and if it can be solved uniquely (or there's some definite guidance), those variables `?1` etc. get unified with the solution we get from Chalk. The way that works is that `Canonicalized::apply_solution` gets called, which correctly handles going back from the canonical solution to the inference context. 

So if you wanted to use the solution you get back here in `callable_sig`, the way to do it would be like here:
https://github.com/rust-analyzer/rust-analyzer/blob/0f7961d5570f17d6c2098ab11d2a3bcbbfb84ff6/crates/ra_hir_ty/src/infer.rs#L331-L337
but as mentioned, that would just be an optimization since it happens automatically later anyway.

flodiebold replied:
To give an example of why this works this way, let's say we have some obligation `SomeType: Fn<(?1,)>`, and an impl like `impl<T> Fn<(Option<T>,)> for SomeType`. So Chalk determines that `?1 = Option<... anything ...>`, but there's no restriction on what the 'anything' can be. The way that looks in the solution is that we get back a `Canonical` with one binder, and the solution is `Option<^0.0>`; so the `^0.0` is a bound variable referring to that binder of the canonical. This means that we need to introduce a new type variable `?2`; and then we have `?1 = Option<?2>` and `?2` can be anything. So "applying the solution" involves potentially learning the values of the original type variables, but also introducing new ones.

adamrk replied:
Ok, I think I've got it now.

adamrk replied:
I made these changes, but a couple other things came up:
1. We still need to run `trait_solve` and make sure that the answer is `Some` so that we pick the right type from autoderef chain. I'm wondering if something could go wrong if we accept ambiguous solutions: is it possible that one type in the autoderef chain doesn't implement `FnOnce`, but `trait_solve` returns `Ambig::Unknown` so we push the wrong obligation? 

1. Actually I guess it's ok to just check the `FnOnce` trait instead of going through all three. I think they need to check all three in rustc because they need more than just the argument and return types.

flodiebold replied:
1. Yeah, but it works the same way in rustc, so it should be fine.
2. Yes. If we do things like move checking we'll need to be more precise, but for now checking for `FnOnce` is actually enough.

flodiebold commented on crates/ra_ide/src/hover.rs:
```diff
@@ -2410,4 +2410,103 @@ fn func(foo: i32) { if true { <|>foo; }; }
+
+            #[lang = "fn"]
+            pub trait Fn<Args> {
+                type Output;
```
this would need to be moved to an `FnOnce` trait

flodiebold commented on crates/ra_ide/src/hover.rs:
```diff
@@ -2410,4 +2410,103 @@ fn func(foo: i32) { if true { <|>foo; }; }
     }
+
+    #[test]
+    fn infer_closure_arg() {
```
oh, I actually missed that these are hover tests. Tests for type inference should mostly be in `ra_hir_ty` :slightly_smiling_face: I'd probably put them in `tests/traits.rs`. While we're at it, we can turn them into snapshot tests to cover both the inferred return value and arguments.

adamrk commented on crates/ra_hir_ty/src/infer/coerce.rs:
```diff
@@ -38,8 +38,8 @@ impl<'a> InferenceContext<'a> {
-                let sig1 = ty1.callable_sig(self.db).expect("FnDef without callable sig");
-                let sig2 = ty2.callable_sig(self.db).expect("FnDef without callable sig");
+                let sig1 = self.callable_sig(ty1).expect("FnDef without callable sig");
+                let sig2 = self.callable_sig(ty2).expect("FnDef without callable sig");
```
Done

adamrk commented on crates/ra_hir_ty/src/infer/coerce.rs:
```diff
@@ -93,7 +93,7 @@ impl<'a> InferenceContext<'a> {
             // `{function_type}` -> `fn()`
             (ty_app!(TypeCtor::FnDef(_)), ty_app!(TypeCtor::FnPtr { .. })) => {
-                match from_ty.callable_sig(self.db) {
+                match self.callable_sig(&from_ty) {
```
Done

adamrk commented on crates/ra_hir_ty/src/infer/expr.rs:
```diff
@@ -692,7 +768,7 @@ impl<'a> InferenceContext<'a> {
         let method_ty = self.insert_type_vars(method_ty);
         self.register_obligations_for_call(&method_ty);
-        let (expected_receiver_ty, param_tys, ret_ty) = match method_ty.callable_sig(self.db) {
+        let (expected_receiver_ty, param_tys, ret_ty) = match self.callable_sig(&method_ty) {
```
Yeah, you're right. The `method_ty` comes from giving a `FunctionId` to `HirDatabase::value_ty` which will always return a `FnDef`. Did it.

adamrk commented on crates/ra_hir_ty/src/traits/builtin.rs:
```diff
@@ -360,7 +360,7 @@ fn super_trait_object_unsize_impl_datum(
 }
 
-fn get_fn_trait(db: &dyn HirDatabase, krate: CrateId, fn_trait: super::FnTrait) -> Option<TraitId> {
+pub fn get_fn_trait(db: &dyn HirDatabase, krate: CrateId, fn_trait: super::FnTrait) -> Option<TraitId> {
```
Done. Seemed like the natural place for in in `traits` is as a method of `FnTrait` and I changed he name to `get_id`.

adamrk commented on crates/ra_hir_ty/src/infer/expr.rs:
```diff
@@ -63,6 +63,75 @@ impl<'a> InferenceContext<'a> {
+                        value: Obligation::Trait(trait_ref),
+                        environment: trait_env,
+                    });
+                if let Some(Solution::Unique(SolutionVariables(solution))) =
```
2. makes sense, but don't we still need to destruct the solution to get the argument types out? For example, if the solution is `Ambig(Guidance::Unknown)` then we won't be able to create the vector of argument types.

flodiebold replied:
Not if we make the query with a tuple already. I.e. Instead of creating one variable and asking for `Fn<?1>`, make one variable per parameter and ask for `Fn<(?1, ?2, ?3, ...)>`. Then we can just return those variables. (By registering the obligation, the solution will be automatically applied to those variables once it is found. We *could* of course already use the solution here if it is unique, but that'd only be an optimization.)

adamrk commented:
I think I've addressed most of the comments except the ones on refactoring how we handle the solution from `trait_solve`.

Also I added a new test to check that the `Output` type is handled properly and it's failing because it shows the output of a `dyn Fn() -> i32` as `FnOnce::Output<dyn Fn<(), Output = i32>, ()>` instead of reducing it to `i32`. So I still need to look into that more.

flodiebold commented:
> Also I added a new test to check that the Output type is handled properly and it's failing because it shows the output of a dyn Fn() -> i32 as FnOnce::Output<dyn Fn<(), Output = i32>, ()> instead of reducing it to i32. So I still need to look into that more.

Try it with `FnOnce` instead. I think that may be an unrelated problem with lowering the associated type binding.

adamrk commented:
> Try it with `FnOnce` instead. I think that may be an unrelated problem with lowering the associated type binding.

It has the same behavior with `FnOnce`.

flodiebold commented:
Ok. Still, that doesn't look like a problem in this PR to me; you could file an issue for it.

flodiebold commented:
Looks good to me!

bors r+
